### PR TITLE
Add support for per-emitter sampling weight

### PR DESCRIPTION
**Pull request by dvicini**
Hi,

I've recently had the need for more customized light sampling (instead of uniform sampling of all available emitters). Out of the box, Mitsuba 3 only supports uniform emitter sampling, which is somewhat limiting, in particular for applications that require on-the-fly modification of light sources. Such modifications are essential for inverse rendering problems involving some form of active illumination (e.g., see also https://github.com/mitsuba-renderer/mitsuba3/discussions/402). Currently the only way to do this seems to be to write a custom integrator. This is always an option, but a relatively heavy solution to just change the sampling of emitters (It effectively requires a duplicating every integrator that one wants to use). I don't think there is another way currently, or at least I couldn't figure out a good way to do it (e.g., a custom emitter plugin that internally handles these sampling decisions doesn't quite work, as we can't attach one emitter to multiple shapes and can't replace a shape's emitter easily after scene loading).

This PR adds a per-emitter `sampling_weight`. The design for this follows Mitsuba 0.6, which also had this weight. The code falls back to the fast uniform sampling if all sampling weights are equal to 1, so that there shouldn't be a performance regression for the default case. I also added some rudimentary tests to `test_scene.py`.

In the future, it could be interesting to factor out all the emitter sampling into a new plugin type (e.g., `EmitterSampler`), to allow users to more selectively replace functionality that's currently hardcoded in `scene.cpp`. But for now I think just re-introducing the sampling weight is probably the easiest solution, which doesn't modify any major interfaces and offers a clear path for users without a deep knowledge of the inner workings of the system. Eventually, introducing a special plugin that handles these sampling decisions could make sense, as it would allow customization at the user level without the need to add much code to Mitsuba itself.

(still slightly work in progress)

### CODE REVIEW

**Comment by njroussel** [comment]
Hi @dvicini 

This is looking good! I don't have any comments right now. Please ping me again once its' fully ready. Or are you finished now?

------

>  Eventually, introducing a special plugin that handles these sampling decisions could make sense, as it would allow customization at the user level without the need to add much code to Mitsuba itself.

We had discussed this at some point and it seemed like a great idea. However, this would come at the cost of an extra virtual function call in scalar mode which we'd like to avoid. I think this was the sole reason why we dropped this idea. I think this could still be something we offer, at least, in JIT variants. For scalar variants, it would at least be nice to know the cost of the extra vcall.

**Comment by dvicini** [comment]
Thanks for taking a look. I will take another pass to make sure it's all correct and then let you know before merging :)

Agree with the undesirable cost for the extra VCall, this is the main downside of that approach. For the JIT variants it's all fine because either way it can be inlined. Maybe there could be some static workaround for the default case when no specialized sampling is used. Either way, I think for now the sampling weight is the best solution due to not creating any significant overhead.

**Comment by dvicini** [comment]
Updated with a mark_dirty flag to tag updated emitters and cleaned up the PR overall. This is ready for review / merge now

File: include/mitsuba/render/emitter.h [hunk 1]
```diff
@@ -74,6 +88,12 @@ class MI_EXPORT_LIB Emitter : public Endpoint<Float, Spectrum> {
 protected:
     /// Combined flags for all properties of this emitter.
     uint32_t m_flags;
+
+    /// Sampling weight
+    ScalarFloat m_sampling_weight;
+
+    /// True if the emitters's parameters have changed
+    bool m_dirty;
```
Comment (njroussel) [inline include/mitsuba/render/emitter.h:96]: Is this ever initialized? Or reset to false after the update?

File: src/render/scene.cpp [hunk 2]
```diff
@@ -334,6 +370,18 @@ MI_VARIANT void Scene<Float, Spectrum>::parameters_changed(const std::vector<std
...
+    // the emitter sampling distribution.
+    bool emitters_changed = false;
+    for (auto &e : m_emitters) {
+        if (e->dirty()) {
+            emitters_changed = true;
+            break;
+        }
+    }
+    if (emitters_changed)
+        update_emitter_sampling_distribution();
```
Comment (njroussel) [inline src/render/scene.cpp:384]: "Unmark" the emitters as dirty here.

**Review by njroussel — CHANGES_REQUESTED** [review]
A couple of comments regarding the `mark_dirty` implementation.

Also, I can't put this comment on the correct line: 
In `point.cpp` the `parameters_changed()` method should also call the `Base::parameters_changed()`.

**Comment by dvicini** [comment]
Good catch! I fixed both point.cpp and the mark_dirty stuff

**Pull request merged**